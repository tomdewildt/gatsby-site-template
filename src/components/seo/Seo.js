/* eslint-disable complexity, max-len */
import React from "react";
import { graphql, useStaticQuery } from "gatsby";
import { Helmet } from "react-helmet";
import PropTypes from "prop-types";

const query = graphql`
    query SEO {
        site {
            siteMetadata {
                siteTitle
                siteLanguage
                siteDescription
                siteBanner
                siteUrl
            }
        }
    }
`;

const Seo = ( {
    title,
    description,
    banner,
    path,
} ) => {
    const { site: { siteMetadata: metadata } } = useStaticQuery( query );

    return (
        <Helmet>
            <html lang={ metadata.siteLanguage } />

            <title>{ title ?? metadata.siteTitle }</title>

            <meta name="title" content={ title ?? metadata.siteTitle } />
            <meta name="description" content={ description ?? metadata.siteDescription } />

            <meta property="og:type" content="website" />
            <meta property="og:url" content={ metadata.siteUrl + path } />
            <meta property="og:title" content={ title ?? metadata.siteTitle } />
            <meta property="og:description" content={ description ?? metadata.siteDescription } />
            <meta property="og:image" content={ banner ?? metadata.siteBanner } />

            <meta property="twitter:card" content="summary_large_image" />
            <meta property="twitter:url" content={ metadata.siteUrl + path } />
            <meta property="twitter:title" content={ title ?? metadata.siteTitle } />
            <meta property="twitter:description" content={ description ?? metadata.siteDescription } />
            <meta property="twitter:image" content={ banner ?? metadata.siteBanner } />
        </Helmet>
    );
};

Seo.propTypes = {
    title: PropTypes.string,
    description: PropTypes.string,
    banner: PropTypes.string,
    path: PropTypes.string,
};

Seo.defaultProps = {
    title: null,
    description: null,
    banner: null,
    path: "",
};

export default Seo;
