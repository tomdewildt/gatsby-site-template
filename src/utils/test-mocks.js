jest.mock( "gatsby", () => ( {
    ...jest.requireActual( "gatsby" ),
    graphql: jest.fn( () => "" ),
    useStaticQuery: jest.fn( () => ( {
        site: {
            siteMetadata: {
                siteTitle: "title",
                siteLanguage: "en",
                siteDescription: "description",
                siteBanner: "https://example.com/banner.png",
                siteUrl: "https://example.com",
            },
        },
    } ) ),
} ) );
