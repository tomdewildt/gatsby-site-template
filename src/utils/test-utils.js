/* eslint-disable import/no-extraneous-dependencies,react/prop-types */
import React from "react";
import { ThemeProvider } from "styled-components";
import { render, waitFor } from "@testing-library/react";

import defaultTheme from "../theme";

export const findByAttribute = async ( key, value ) => {
    let result = null;
    await waitFor( () => {
        result = document.querySelector( `[${ key }='${ value }']` );
        return expect( result ).toBeInTheDocument();
    } );
    return result;
};

export const findAllByAttribute = async ( key, value ) => {
    let result = null;
    await waitFor( () => {
        result = document.querySelectorAll( `[${ key }='${ value }']` );
        return expect( result.length ).toBeGreaterThan( 0 );
    } );
    return result;
};

export const renderWithTheme = (
    component,
    theme = defaultTheme,
) => ( {
    findByAttribute,
    findAllByAttribute,
    ...render( component, {
        wrapper: ( { children } ) => (
            <ThemeProvider theme={ theme }>
                { children }
            </ThemeProvider>
        ),
    } ),
} );

export const renderComponent = (
    component,
    options = {},
) => ( {
    findByAttribute,
    findAllByAttribute,
    ...render( component, options ),
} );
