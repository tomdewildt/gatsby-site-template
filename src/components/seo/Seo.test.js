import React from "react";
import { axe } from "jest-axe";

import { renderWithTheme } from "../../utils/test-utils";

import Seo from "./Seo";

describe( "Seo", () => {
    it( "renders consistently", () => {
        const { asFragment } = renderWithTheme( <Seo /> );
        const fragment = asFragment();

        expect( fragment ).toMatchSnapshot();
    } );

    it( "should have no axe violations", async () => {
        const { container } = renderWithTheme( <Seo /> );
        const result = await axe( container );

        expect( result ).toHaveNoViolations();
    } );

    it( "respects the title prop", async () => {
        const { findByAttribute } = renderWithTheme( <Seo title="custom title" /> );
        const metaTitle = await findByAttribute( "name", "title" );
        const ogTitle = await findByAttribute( "property", "og:title" );
        const twitterTitle = await findByAttribute( "property", "twitter:title" );

        expect( document.title ).toEqual( "custom title" );
        expect( metaTitle.content ).toEqual( "custom title" );
        expect( ogTitle.content ).toEqual( "custom title" );
        expect( twitterTitle.content ).toEqual( "custom title" );
    } );

    it( "respects the description prop", async () => {
        const { findByAttribute } = renderWithTheme( <Seo description="custom description" /> );
        const metaDescription = await findByAttribute( "name", "description" );
        const ogDescription = await findByAttribute( "property", "og:description" );
        const twitterDescription = await findByAttribute( "property", "twitter:description" );

        expect( metaDescription.content ).toEqual( "custom description" );
        expect( ogDescription.content ).toEqual( "custom description" );
        expect( twitterDescription.content ).toEqual( "custom description" );
    } );

    it( "respects the banner prop", async () => {
        const { findByAttribute } = renderWithTheme( (
            <Seo banner="https://example.com/images/custom.png" />
        ) );
        const ogBanner = await findByAttribute( "property", "og:image" );
        const twitterBanner = await findByAttribute( "property", "twitter:image" );

        expect( ogBanner.content ).toEqual( "https://example.com/images/custom.png" );
        expect( twitterBanner.content ).toEqual( "https://example.com/images/custom.png" );
    } );

    it( "respects the path prop", async () => {
        const { findByAttribute } = renderWithTheme( <Seo path="/custom" /> );
        const ogUrl = await findByAttribute( "property", "og:url" );
        const twitterUrl = await findByAttribute( "property", "twitter:url" );

        expect( ogUrl.content ).toEqual( "https://example.com/custom" );
        expect( twitterUrl.content ).toEqual( "https://example.com/custom" );
    } );
} );
