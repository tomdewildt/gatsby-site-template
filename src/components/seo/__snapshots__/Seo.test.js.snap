// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Seo renders consistently 1`] = `<DocumentFragment />`;
